### ISSUE TITLE
Proposal for a new type of refund

### ISSUE THREAD

**Issue by hackaugusto** [raiden-network/raiden-contracts#1216]
This is a rough sketch, it is by no means correct. I'm opening the issue so that we don't lose this proposal:

```
All values bellow are monotonic increasing.

## Settlement computation on-chain

Node's values set on the smart contract:
D = confirmed deposit
W = withdraw

Partner's balance proof values sent by the node:
L = locked
U = unlocked
E = expired
R = refund

Node's balance proof values sent by the partner:
R' = refund
U' = unlock

Computed
N = net
P = pending

Constraints:
L,U,E,R,R',U',D,W >= 0
L >= U + E

Formulas:

N = D - W - U' + U
P = L - max(R', E) - U
```

The idea here is: For a channel A-B, the above definition would allow B to send a balance proof that changes the locked amount of A, with this balance proof A could prove to the smart contract that the tokens have been returned by B to A, so no backwards transfer would be required.

The biggest question that I have with the above is that if a locksroot is also required, or if just a returned amount is sufficient.

**Comment by pirapira** [comment]
`max` in `max(R', E)` looks weird to me.  You don't care about the value of `R'` when you have lots of expired locks?

**Comment by pirapira** [comment]
> The biggest question that I have with the above is that if a locksroot is also required, or if just a returned amount is sufficient.

I think a locksroot is required.  Say two transfers with two different secret hashes were refunded.  And say only one of the secrets has been revealed.  Now it matters a lot which lock was refunded and which lock is alive.

**Comment by hackaugusto** [comment]
> max in max(R', E) looks weird to me. You don't care about the value of R' when you have lots of expired locks?

I don't understand. How does `max(R', E)` mean that expired locks are ignored? (I'm assuming that is what you meant by don't care).

The core idea here is that both participants will be changing the same value. To avoid synchronization among the parties the value is made monotonic and `max` is used, otherwise we would require some sort of synchronization among the nodes to change this value. The next thing is to make sure the result of `max(R', E)` is semantically correct and not exploitable.

**Comment by pirapira** [comment]
`max(0,100)` is equal to `max(20,100)`.  So when `E` is `100`, you don't care whether `R'` is `0` or `20`.

**Comment by pirapira** [comment]
That's weird.  When the channel is very young, `E` is still zero, so every change in `R'` matters.  When the channel is very old after many expired locks of some value, `E` would be big, and you don't care about small `R'`s anymore.

I think `E` can grow big while `R'` can stay at zero.

**Comment by hackaugusto** [comment]
> max(0,100) is equal to max(20,100). So when E is 100, you don't care whether R' is 0 or 20.

Yes, that is intentional.

It has to possibilities, `R' < E` or `R' > E `. The value of `R'` is controlled by the partner node, while the value of `E` is controlled by the local node. For the case `R' < E` the result would be `E`, this means the local node does not have to wait for the partner to send a `RemovedExpiredLock` message. For the case `E < R'` is the opposite, the partner does not have to wait for the local node to send a `RefundTransfer`

**Comment by pirapira** [comment]
@hackaugusto 

1. I send locked transfers but never reveal secrets.  So `E == 100` or something.

2. I send a payment through this channel but the routing fails, so I get a refund `R' == 20`.

3. The max computation kills my refund.

**Comment by hackaugusto** [comment]
@pirapira That would be an invalid message, it does not refund anything, the correct value for `R'` is the previous `E` in addition to the new refund amount, so it should be `120`.

**Comment by pirapira** [comment]
Refunds include expired locks?

**Comment by hackaugusto** [comment]
With this definition, yes. Pretty much a refund message would be partner node saying it's okay for the local node to use the tokens from a given lock, because the partner is "anticipating" the expiration.

**Comment by pirapira** [comment]
I see definitions:
```
R' = refund
E = expired
```
but  I see no equations between `refund` and `expired`.

**Comment by pirapira** [comment]
Are they somehow related?

**Comment by pirapira** [comment]
Is it intentional that

```
U' = unlock
```

Has `unlock` instead of `unlocked`?

**Comment by hackaugusto** [comment]
> Are they somehow related?

Yes, they are "the same value" from the settlement algorithm perspective. This happens because the `max` is used in formula `P = L - max(R', E) - U`.

**Comment by pirapira** [comment]
So, `R'` and `E` supposed to be equal in a completely synchronous world?

**Comment by pirapira** [comment]
Still `max` is weird.

Let's say we have an event 1 that inceases `R'`.  And event 2 that increases `E`.  Both events increase the value by 10.  In total the increase should be 20.

Event 1 happened on the peer's side so I see `R' == 10`.  Event 2 happened on my side so I see `E == 10`.  These two ten's come from different causes so the perfectly synchronous world I should see 20 already, but the `max` calculation only gives me 10.

**Comment by pirapira** [comment]
> Yes, they are "the same value" from the settlement algorithm perspective. This happens because the max is used in formula P = L - max(R', E) - U.

Now you've put me in a loop. I asked why `max` and you answered it's because the same value.  It's the same value because the max is used?

**Comment by pirapira** [comment]
I need to see what kind of event increases which variable.

**Comment by hackaugusto** [comment]
Here are two examples:

A -> B; LockedTransfer; locked_amount = 10
B -> A; Refund; refund_amount = 10

With the above, there is nothing pending, so `P = L - max(R', E) - U` would be `P = 0 - max(10, 0) - 0`, and there is not transfer of value. Which means the partner did a refund without locking additional funds.

Let's add some interaction:

A -> B; LockedTransfer; locked_amount = 10
A -> B; LockedTransfer; locked_amount = 20
B -> A; Refund; refund_amount = 10
A -> B; Unlock; unlocked_amount = 10

In this case, there is nothing pending `P = L - max(R', E) - U` would be `P = 20 - max(10, 0) - 10`, and the partner does get the first transfer, because it would net 10 tokens, from the formula `N = D - W - U' + U` would be `N = D - 0 - 0 + 10` (Note: I had to fix the net formula)

**Comment by hackaugusto** [comment]
This is a hard case for me:

1. A->B; LockedTransfer
1. A->B; LockedTransfer
1. Time passes
1. B->A; RefundTransfer for 1
1. A->B; RemoveExpiredLock for 2

I don't really know how to fix that race.

**Comment by pirapira** [comment]
> A -> B; LockedTransfer; locked_amount = 10
> A -> B; LockedTransfer; locked_amount = 20
> B -> A; Refund; refund_amount = 10
> A -> B; Unlock; unlocked_amount = 10

So the two parties disagree on the first lock, whether it's refunded or unlocked?

**Comment by pirapira** [comment]
When you send RemoveExpiredLock, do you write a bigger `E`?

When you receive Refund, do you expect a bigger `R'`?

**Comment by pirapira** [comment]
```
1. A->B; LockedTransfer
2. A->B; LockedTransfer
3. Time passes
4. B->A; RefundTransfer for 1
5. A->B; RemoveExpiredLock for 2
```

What's hard about this?  I see no conflicting information.

**Comment by hackaugusto** [comment]
> So the two parties disagree on the first lock, whether it's refunded or unlocked?

No, the values are monotonic, so a locked_amount means a second lock of `10`, I didn't specify which lock was unlocked an which was refunded in the example.

**Comment by hackaugusto** [comment]
> When you send RemoveExpiredLock, do you write a bigger E?
> When you receive Refund, do you expect a bigger R'?

Yes

**Comment by pirapira** [comment]
Then why do `E` and `R'` grow together?

**Comment by hackaugusto** [comment]
> Then why do E and R' grow together?

They are changed by different computers separated by a network in between.

**Comment by pirapira** [comment]
I heard `E` and `R'` are "the same value" (in the synchronous world).

**Comment by hackaugusto** [comment]
> What's hard about this? I see no conflicting information.

One of the messages will be invalid, and then we have to reprocess the queue of messages. The whole point of using monotonically increasing values is to avoid synchronization, we need a policy on how to move forward even if that happens without having to synchronize.

**Comment by pirapira** [comment]
Is there a rule like, "When you receive Refund, you have to send RemoveExpiredLock"?

**Comment by pirapira** [comment]
There has to be some connection between `E` and `R'`?

**Comment by pirapira** [comment]
```
1. A->B; LockedTransfer
2. A->B; LockedTransfer
3. Time passes
4. B->A; RefundTransfer for 1
5. A->B; RemoveExpiredLock for 2
```

I can see all messages being valid.  What kind of assumption am I missing?

**Comment by pirapira** [comment]
Maybe the locks must be removed in the same order as they are created?

**Comment by pirapira** [comment]
Maybe it's easier to start with the lifecycle of a hashtime lock.

So a hashtime lock can have these states.
a. hasn't been created
b. A added it in A-->B side of the channel state
c. B saw it in A-->B side of the channel state
d. expired
e. unlocked
f. refunded

And a lock can move
* from a to b: L increases
* from b to c: no change on A's local view
* from c to d: E increases
* from c to e: U increases
* from c to f: R' increases
* from b to d: E increases

Right?

**Comment by hackaugusto** [comment]
> Is there a rule like, "When you receive Refund, you have to send RemoveExpiredLock"?

Unless the current rules don't work (as in, send the remove expired lock after the lock expires), I wouldn't change it. But I did not think it in detail to see if there is a corner case that breaks this proposal.

> There has to be some connection between E and R'?

Yes:

- Either must be preceded by a locked transfer from the local node. A lock that doesn't exist cannot be expired nor refunded.
- The values are used in the formula mentioned previously.

Effectively the defined values are saying that for the life time of the channel "X tokens were locked in this direction" `locked_amount`. "X tokens were moved to the partner" `unlocked_amount`. "X tokens were returned because the transfer failed" (`refund_amount` and `expired_amount`).

> Maybe the locks must be removed in the same order as they are created?

That would be bad for refunds, either only the first or last lock would be refundable.

**Comment by pirapira** [comment]
> "X tokens were returned because the transfer failed" (refund_amount and expired_amount).

Isn't `expired_amount` about the current block number moving ahead and the expiration block has passed?

**Comment by pirapira** [comment]
Maybe refund and expiration are two ways that the transaction fails?

Then I'm imagining now: `E` is about the expired locks from me to the peer. `R'` is about the refunded locks from me to the peer.

**Comment by hackaugusto** [comment]
> Isn't expired_amount about the current block number moving ahead and the expiration block has passed?

The block being larger than the lock expiration is a requirement for this from the sender perspective. But the value itself only increases when the sender sends a balance proof with it. (I have the feeling I didn't get what you meant with this one)

**Comment by pirapira** [comment]
But the expired locks are counted towards the sender's balance, so `E` should be about the locks from the peer to me, that has expired.

**Comment by hackaugusto** [comment]
> Maybe refund and expiration are two ways that the transaction fails?
> Then I'm imagining now: E is about the expired locks from me to the peer. R' is about the refunded locks from me to the peer.

yes

**Comment by pirapira** [comment]
If `E` and `R'` are about locks in different directions, why are they financially the same value?

**Comment by pirapira** [comment]
I miss the table with :arrow_up: and :arrow_down: filled for all messages and all variables.

**Comment by hackaugusto** [comment]
> If E and R' are about locks in different directions, why are they financially the same value?

expired -> used by the sender
refund -> used by the receiver

They both have the end effect of returning locked tokens to the sender.

> I miss the table with arrow_up and arrow_down filled for all messages and all variables.

All values are monotonically increasing, what changes up or down is the balance (which is computed from the monotonic values)

**Comment by pirapira** [comment]
I still don't get it.

* I send a lock A to the peer.  This lock expires.  That's counted in `E`.
* The peer sends a lock B to me.  I refund this lock.  That's counted in `R``.

Why do these values need to be equal?

**Comment by pirapira** [comment]
Why do I take the `max` of the amount of lock A and the amount of lock B?

**Comment by hackaugusto** [comment]
> Why do these values need to be equal?

In you example they are not, the values are mixed. This would make sense:

I send a lock A to the peer. This lock expires. That's counted in E.
I send a lock B to the peer. The peer refunds this lock. That's counted in R'.

**Comment by pirapira** [comment]
I'm confused.  You wrote

> yes (edit: almost, it should be R' is about the refunded locks from the peer to me.)

So I thought `R'` is about a lock that the peer sends.

**Comment by pirapira** [comment]
> I send a lock A to the peer. This lock expires. That's counted in E.

Now the lock is expired, and removed from the state.  So a refund cannot happen, right?

**Comment by hackaugusto** [comment]
> So I thought R' is about a lock that the peer sends.

Sorry, I misinterpreted your sentence. I read  `R' is about the refunded locks from me to the peer.` as `R' is about the refund_amount for the lock from my peer to me.` and not as `R' how much my peer refunded me from my own locks.`

**Comment by pirapira** [comment]
I see, the message direction vs lock direction.

**Comment by pirapira** [comment]
Or is it the case, the same event looks from one side "expire" and from the other side "refund"?

**Comment by pirapira** [comment]
I thought a lock will be
* unlocked
* refunded
* or becomes expired.

One of the three.

**Comment by hackaugusto** [comment]
> Or is it the case, the same event looks from one side "expire" and from the other side "refund"?

https://github.com/raiden-network/raiden-contracts/issues/1216#issuecomment-528022247

Values with a prime are controlled by the partner, so the message with the balance proof comes from the partner. Values with **no prime** are controlled by the local node, so the balance proof goes from the local node to the partner.

Values from both directions are used to compute how many tokens are still locked, what is the finalized transferred amount, and what is the pending amount.

**Comment by pirapira** [comment]
I control `E` and my peer controls `R'`.  `E` is about the locks I sent that has expired (block number).  `R'` is about the locks I sent that has been refunded (routing failure).  So if the peer is nice, he refunds often `R'` tends to go up.  If the peer is not nice, he leaves locks till they expire, so `E` tends to go up.  Right?

**Comment by pirapira** [comment]
I'd add `E` and `R'` then.

**Comment by hackaugusto** [comment]
> I'd add E and R' then.

what about the race I mentioned above?

**Comment by pirapira** [comment]
It works perfectly.

**Comment by pirapira** [comment]
1. A->B; LockedTransfer
2. A->B; LockedTransfer
3. Time passes
4. B->A; RefundTransfer for 1
5. A->B; RemoveExpiredLock for 2

4. increases `R'`.  and 5. increases `E`.  I add them.

**Comment by hackaugusto** [comment]
Oopps, I had a different example in mind:

1. A->B; LockedTransfer
1. Time passes
1. B->A; RefundTransfer
1. A->B; RemoveExpiredLock

**Comment by pirapira** [comment]
I see the problem.  Addition `E + R'` is wrong (can be too big).  `max` is not a solution either (can be too small).

I believe `E` is better accounted in the other side of the channel.

Currently it's like "I sent you a lock, but believe me, it's expired, so forget about the lock."

In the fixed version, "I received a lock from you, but I saw blocks and the secret hasn't been registered.  I'm sure this lock is useless.  So let's forget about the lock."

Then `E` and `R` will be accounted in the same side (or even be merged into a single variable).

Raiden has a principle that every bigger number in a balance proof must be good for the receiver of the balance proof.  In the current protocol, bigger `E` is not better for the receiver of the balance proof.  In the fixed version, a bigger `E` is better for the receiver of the balance proof (who is the sender of the lock, and after seeing a bigger `E`, she can get rid of the obligation).

**Comment by pirapira** [comment]
Up'ping the priority because @kilrau says it's important for them https://github.com/raiden-network/raiden/issues/4298#issuecomment-537001222 .
